### PR TITLE
Extend CircleCI integration

### PR DESCRIPTION
**Pull request by lrnt**
This PR extends the CircleCI pipeline in separate jobs to:

1. bootstrap the monorepo
2. run the library and contract tests
3. generate the documentation if the tests are successful

The CircleCI jobs and workflow in action can be seen here: https://circleci.com/gh/lrnt/requestNetwork

image

There are some things that could be improved/added in the future:

- [ ] Use a Docker image with Node which comes preinstalled with Lerna to speedup each job
- [X] Upload the documentation to S3 like what is currently happening with Travis
- [ ] Add a build status badge to README.md

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,31 +1,116 @@
...
+    working_directory: *working_directory
+    steps:
+      - attach_workspace:
+          at: *working_directory
+      - run:
+          name: 'Install lerna'
+          command: 'sudo npm install -g lerna'
+      - run:
+          name: 'Deploy test contract'
+          command: 'lerna run --scope @requestnetwork/request-network.js testdeploy'
```
Comment (romaric-juniet) [inline .circleci/config.yml]: Tests for request-network-smart-contracts don't require to deploy the contracts with `testdeploy`

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,31 +1,116 @@
...
+          command: 'lerna run --scope request-network-smart-contracts test --stream'
+  docs-library:
+    docker:
+      - *node_image
+    working_directory: *working_directory
+    steps:
+      - attach_workspace:
+          at: *working_directory
+      - run:
+          name: 'Install lerna'
```
Comment (romaric-juniet) [inline .circleci/config.yml]: Description could be "Install lerna and compodoc"

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,31 +1,116 @@
...
+          command: 'sudo npm install -g lerna'
+      - run:
+          name: 'Deploy test contract'
+          command: 'lerna run --scope @requestnetwork/request-network.js testdeploy'
+      - run:
+          name: 'Test requestNetwork.js library'
+          command: 'lerna run --scope @requestnetwork/request-network.js test --stream'
```
Comment (romaric-juniet) [inline .circleci/config.yml]: the `--stream` option was to have the console output in travis. It makes the logs a bit harder to read though, maybe it's not needed for circleCI

File: .circleci/config.yml [hunk 4]
```diff
@@ -1,31 +1,116 @@
...
+    steps:
+      - attach_workspace:
+          at: *working_directory
+      - run:
+          name: 'Install lerna'
+          command: 'sudo npm install -g lerna @compodoc/compodoc'
+      - run:
+          name: 'Create library documentation'
+          command: 'lerna run --scope @requestnetwork/request-network.js docs --stream'
+  docs-contract:
```
Comment (romaric-juniet) [inline .circleci/config.yml]: I notice the smart contract linting is missing

**Comment by lrnt** [comment]
Smart contract linting and S3 documentation deployment has been added. For the S3 deployment to work the following environment variables need to be configured in the requestNetwork CircleCI project:

- `AWS_ACCESS_KEY_ID`
- `AWS_SECRET_ACCESS_KEY`

**Comment by romaric-juniet** [comment]
Looks awesome Laurent, thanks for working on this. I'll finish the configuration with the AWS keys and let you know when it's all done

**Pull request merged**